great_expectations/profiler/rule/rule_state.py

```diff
@@ -20,22 +21,21 @@ class RuleState:
 
     def __init__(
         self,
-        active_domain: Optional[
-            Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]
-        ] = None,
-        domains: Optional[
-            List[Dict[str, Union[str, MetricDomainTypes, Dict[str, Any]]]]
-        ] = None,
+        active_domain: Optional[Domain],
```

Comment:
Do we need this distinction between domains and active_domains for a Rule State? Is it because we would want to deactivate a domain via user request? I think this is a good idea -  but just verifying.

Reply:
@gilpasternak35 This is a great question.  The answer is: "I am not sure yet".  It all comes from the loop starting with the line `for domain in rule_state.domains:` in `rule.py`.  Maybe `rule_state.py` does not need `domains` and only the `active_domain`?  I do not see `rule_state.py` making use of its `domains` state variable.  However, `active_domain` is used in important capacity by `rule_state.py` (in `get_parameter_value()`).  Furthermore, it is import that `active_domain` remain a state variable of `rule_state.py`, because `ExpectationConfigurationBuilder` uses it: `rule_state.get_parameter_value()`.  But `rule.py` itself does not maintain domains -- it retrieves them from `rule_state.py`, which means something has to set `domains` on `RuleState`.  Hence, we either keep it this way, or we remove `domains` from `RuleState` and maintain them in another place.  It all depends on how the rule is built.  Thoughts welcome, and I will also be thinking about it, and we should all discuss this.  Thank you!

Reply:
@alexsherstinsky this makes a lot of sense to me. I think this leads us to the fundamental question here: is a rule a certain type of validation proposed on a certain type of domain, or is a rule just a certain type of validation and rule_state a way to propose all domains upon which the validation should be applied? Given what I've seen in the codebase and our discussion here, I lean strongly towards the latter, but I still feel that this is a distinction that should be clearly made (please forgive me if this is already clearly specified in our concepts doc and I just missed it). If we use the latter definition, then I think it is intuitive to maintain a list of domains in rule_state

Reply:
Great conversation! The way I have been thinking about this is that a rule is the blueprint for an expectation with all of its necessary metrics. We loop through all of the domains returned by the `DomainBuilder`, retrieve the appropriate metrics and store them in the `rule_state`. The `ExpectationConfigurationBuilder` then uses the `rule_state` to add the defined expectations to our suite.

Based on this (which is just my expectation of how this would work, and not how it actually works), I wouldn't really see the need for an `active_domain`. I would instead expect that `get_parameter_value()` would be a method on the `ExpectationConfigurationBuilder`. Otherwise, as we build `ExpectationConfigurations` for all of the domains returned by the `DomainBuilder`, we will need to modify the `active_domain` of the `rule_state` each time in order to use that to return parameters stored elsewhere in the `rule_state`, which feels like an odd pattern.

Reply:
@talagluck and @gilpasternak35 Thank you for your inputs -- I actually see something valuable in both of your perspectives.  Since incorporating these ideas would exceed the scope of the present PR, how about we merge this one (barring any show-stoppers), while keeping this comment thread open, to be addressed in the very near future?  Thank you!

Reply:
This all makes sense. We haven't discussed this much yet (just briefly), but what about expectations that reference other domains e.g. A, B, C domains and I want to specify the min_value of my `expect_column_max_to_be_between` for all domains to be the max of domain A. There are probably other ways to do this, but in this case it may be helpful to have the concept of an `active_domain` since there will be two domain references floating around (domain A and the currently `active_domain` as we are looping through). Just something to keep in mind, I'm not sure yet how we will implement this.

Reply:
Totally agree @alexsherstinsky !

Reply:
Ah - that's a great point, @anthonyburdi! It would be great to brainstorm that out since it sounds like a pretty important case (how can we stored linked/secondary domains)?. That said, I would still imagine that something like `active_domain` would be in the ExpectationConfigurationBuilder rather than `rule_state`,

Reply:
@anthonyburdi I agree with your observations, too.  This just says very clearly that we need to first and foremost update our concept document and any diagrams, and then get into a complete agreement of the separation of concerns between rule, rule state, and expectation configuration builder.  Further, we have nothing on rule builder as of yet (if such an entity makes sense), though I am assuming that it would mainly reflect the configuration.  I would love to get clarity on all of these things by some time tomorrow.

Reply:
That makes sense @talagluck - and it also may be out of scope for V1 but yes let's absolutely discuss